### PR TITLE
gh-62432: unittest runner: Exit code 5 if no tests were run

### LINKED ISSUE #62432: running a suite with no tests is not an error
BPO | [18232](https://bugs.python.org/issue18232)
--- | :---
Nosy | @terryjreedy, @rbtcollins, @ezio-melotti, @voidspace, @mgorny, @kamilturek
PRs | <li>python/cpython#24893</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2013-06-16.19:32:39.867>
labels = ['type-feature', 'tests', '3.10']
title = 'running a suite with no tests is not an error'
updated_at = <Date 2021-03-16.12:25:51.548>
user = 'https://github.com/rbtcollins'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-16.12:25:51.548>
actor = 'mgorny'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Tests']
creation = <Date 2013-06-16.19:32:39.867>
creator = 'rbcollins'
dependencies = []
files = []
hgrepos = []
issue_num = 18232
keywords = ['patch']
message_count = 8.0
messages = ['191282', '191611', '192331', '226617', '226620', '226631', '388669', '388682']
nosy_count = 6.0
nosy_names = ['terry.reedy', 'rbcollins', 'ezio.melotti', 'michael.foord', 'mgorny', 'kamilturek']
pr_nums = ['24893']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue18232'
versions = ['Python 3.10']
```

</p></details>

### Linked PRs
* gh-102051

### PR DESCRIPTION
**Pull request by stefanor**
As discussed in https://discuss.python.org/t/unittest-fail-if-zero-tests-were-discovered/21498/7

It is common for test runner misconfiguration to fail to find any tests,
this should be an error.

Fixes: #62432

* Issue: gh-62432

### CODE REVIEW

**Comment by terryjreedy** [comment]
Tests/macOS is routinely failing; ignore currently.

This is a design bug fix, hence enhancement (feature change), as much as I would like to pretend otherwise.

File: Lib/unittest/main.py [hunk 1]
```diff
@@ -273,6 +273,11 @@ def runTests(self):
             testRunner = self.testRunner
         self.result = testRunner.run(self.test)
         if self.exit:
-            sys.exit(not self.result.wasSuccessful())
+            if self.result.testsRun == 0:
+                sys.exit(5)
+            elif not self.result.wasSuccessful():
+                sys.exit(1)
+            else:
+                sys.exit(0)
```
Comment (hugovk) [inline Lib/unittest/main.py]: Let's either flip these so positive comes first:
```suggestion
            elif self.result.wasSuccessful():
                sys.exit(0)
            else:
                sys.exit(1)

```

Or keep the original one-liner:

```suggestion
            else:
                sys.exit(not self.result.wasSuccessful())

```

(Plus an intentional newline at the end so we have 2 blank lines after the class.

File: Lib/unittest/runner.py [hunk 2]
```diff
@@ -238,6 +238,8 @@ def run(self, test):
                 infos.append("failures=%d" % failed)
             if errored:
                 infos.append("errors=%d" % errored)
+        elif run == 0:
+            self.stream.write("NO TESTS RUN")
```
Comment (hugovk) [inline Lib/unittest/runner.py]: Also matches tense of pytest:
```suggestion
            self.stream.write("NO TESTS RAN")
```


```console
$ pytest
============================================================================ test session starts =============================================================================
platform darwin -- Python 3.11.2, pytest-7.2.1, pluggy-1.0.0
rootdir: /Users/huvankem/github/norwegianblue, configfile: pyproject.toml
plugins: forked-1.4.0, xdist-3.2.0, timeout-2.1.0, respx-0.20.1, cov-4.0.0, repeat-0.9.1, anyio-3.6.2, lazy-fixture-0.6.3, hypothesis-5.49.0
collected 0 items

=========================================================================== no tests ran in 0.02s ============================================================================
```

File: Lib/test/test_unittest/test_program.py [hunk 3]
```diff
@@ -71,15 +71,22 @@ def testExpectedFailure(self):
         def testUnexpectedSuccess(self):
             pass
 
-    class FooBarLoader(unittest.TestLoader):
-        """Test loader that returns a suite containing FooBar."""
+    class Empty(unittest.TestCase):
+        pass
+
+    class TestLoader(unittest.TestLoader):
+        """Test loader that returns a suite containing testsuite."""
```
Comment (hugovk) [inline Lib/test/test_unittest/test_program.py]: ?

```suggestion
        """Test loader that returns a suite containing self.testcase."""
```

File: Lib/test/test_unittest/test_runner.py [hunk 4]
```diff
@@ -574,6 +574,16 @@ def test(self):
...
 
+    def test_run_empty_suite_error_message(self):
+        class EmptyTest(unittest.TestCase):
+            pass
+
+        suite = unittest.defaultTestLoader.loadTestsFromTestCase(EmptyTest)
+        runner = getRunner()
+        runner.run(suite)
+
+        self.assertIn("\nNO TESTS RUN\n", runner.stream.getvalue())
```
Comment (hugovk) [inline Lib/test/test_unittest/test_runner.py]:
```suggestion
        self.assertIn("\nNO TESTS RAN\n", runner.stream.getvalue())
```

File: Misc/NEWS.d/next/Library/2023-02-19-12-37-08.gh-issue-62432.GnBFIB.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+The :mod:`unittest` runner will now exit with status code 5, if no tests
```
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2023-02-19-12-37-08.gh-issue-62432.GnBFIB.rst]:
```suggestion
The :mod:`unittest` runner will now exit with status code 5 if no tests
```

File: Lib/test/test_unittest/test_program.py [hunk 6]
```diff
@@ -161,6 +167,17 @@ def test_ExitAsDefault(self):
...
+    def test_ExitEmptySuite(self):
+        stream = BufferedWriter()
+        with self.assertRaises(SystemExit) as cm:
+            unittest.main(
+                argv=["empty"],
+                testRunner=unittest.TextTestRunner(stream=stream),
+                testLoader=self.TestLoader(self.Empty))
+        self.assertEqual(cm.exception.code, 5)
+        out = stream.getvalue()
+        self.assertIn('\nNO TESTS RUN\n', out)
```
Comment (hugovk) [inline Lib/test/test_unittest/test_program.py]:
```suggestion
        self.assertIn('\nNO TESTS RAN\n', out)
```

File: Lib/test/test_unittest/test_program.py (hunk 3 above)
Comment (stefanor) [reply Lib/test/test_unittest/test_program.py]: That was a typo of the class initialization argument name, thanks

**Comment by stefanor** [comment]
> @[stefanor](https://github.com/stefanor) stefanor requested review from [hugovk](https://github.com/hugovk) and removed request for [gpshead](https://github.com/gpshead) and [ezio-melotti](https://github.com/ezio-melotti)

Urgh, that removal wasn't intentional, I just hit the re-request review button.

**Comment by stefanor** [comment]
@gpshead: Ping? (it's been 2 months)

**Review by gpshead — APPROVED** [review]

**Comment by stefanor** [comment]
Thank you!

**Comment by vstinner** [comment]
The new Python 3.12 message surprises users when all tests are skipped: see issue https://github.com/python/cpython/issues/113661.

**Comment by fishy** [comment]
The PR description says:

>It is common for test runner misconfiguration to fail to find any tests

Which seems to suggest that it should not fail test if all tests are _skipped_, as skipped tests are still "found", just intentionally skipped from running.

**Comment by stefanor** [comment]
> Which seems to suggest that it should not fail test if all tests are skipped, as skipped tests are still "found", just intentionally skipped from running.

Yeah, I'd agree with that.

**Pull request merged**